File: packages/go-sdk/fsm/step.go
package fsm

import "fmt"

// StepRunStatus represents the status of a workflow step run.
type StepRunStatus string

const (
	StepRunPending   StepRunStatus = "pending"
	StepRunWaiting   StepRunStatus = "waiting"
	StepRunRunning   StepRunStatus = "running"
	StepRunCompleted StepRunStatus = "completed"
	StepRunFailed    StepRunStatus = "failed"
	StepRunSkipped   StepRunStatus = "skipped"
	StepRunCanceled  StepRunStatus = "canceled"
)

// StepRunEvent represents an event that triggers a step run state transition.
type StepRunEvent string

const (
	StepRunEventWait     StepRunEvent = "WAIT"
	StepRunEventStart    StepRunEvent = "START"
	StepRunEventComplete StepRunEvent = "COMPLETE"
	StepRunEventFail     StepRunEvent = "FAIL"
	StepRunEventSkip     StepRunEvent = "SKIP"
	StepRunEventCancel   StepRunEvent = "CANCEL"
)

var stepRunTransitions = map[StepRunStatus]map[StepRunEvent]StepRunStatus{
	StepRunPending: {
		StepRunEventWait:     StepRunWaiting,
		StepRunEventStart:    StepRunRunning,
		StepRunEventSkip:     StepRunSkipped,
		StepRunEventCancel:   StepRunCanceled,
		StepRunEventComplete: StepRunCompleted,
	},
	StepRunWaiting: {
		StepRunEventStart:    StepRunRunning,
		StepRunEventSkip:     StepRunSkipped,
		StepRunEventCancel:   StepRunCanceled,
		StepRunEventComplete: StepRunCompleted,
	},
	StepRunRunning: {
		StepRunEventComplete: StepRunCompleted,
		StepRunEventFail:     StepRunFailed,
		StepRunEventCancel:   StepRunCanceled,
	},
}

var terminalStepRunStatuses = map[StepRunStatus]bool{
	StepRunCompleted: true,
	StepRunFailed:    true,
	StepRunSkipped:   true,
	StepRunCanceled:  true,
}

// CanTransitionStepRun checks whether a step run transition is valid.
func CanTransitionStepRun(from StepRunStatus, event StepRunEvent) bool {
	events, ok := stepRunTransitions[from]
	if !ok {
		return false
	}
	_, ok = events[event]
	return ok
}

// TransitionStepRun returns the next status after applying the event.
func TransitionStepRun(from StepRunStatus, event StepRunEvent) (StepRunStatus, error) {
	events, ok := stepRunTransitions[from]
	if !ok {
		return "", fmt.Errorf("invalid step run status %q", from)
	}
	next, ok := events[event]
	if !ok {
		return "", fmt.Errorf("invalid transition: %q + %q", from, event)
	}
	return next, nil
}

// IsTerminalStepRunStatus returns true if the status is a terminal (final) state.
func IsTerminalStepRunStatus(status StepRunStatus) bool {
	return terminalStepRunStatuses[status]
}

File: packages/ruby-sdk/lib/strait/fsm/step.rb
# frozen_string_literal: true

module Strait
  module FSM
    STEP_RUN_PENDING = "pending"
    STEP_RUN_WAITING = "waiting"
    STEP_RUN_RUNNING = "running"
    STEP_RUN_COMPLETED = "completed"
    STEP_RUN_FAILED = "failed"
    STEP_RUN_SKIPPED = "skipped"
    STEP_RUN_CANCELED = "canceled"

    STEP_RUN_EVENT_WAIT = "WAIT"
    STEP_RUN_EVENT_START = "START"
    STEP_RUN_EVENT_COMPLETE = "COMPLETE"
    STEP_RUN_EVENT_FAIL = "FAIL"
    STEP_RUN_EVENT_SKIP = "SKIP"
    STEP_RUN_EVENT_CANCEL = "CANCEL"

    TERMINAL_STEP_RUN_STATUSES = [
      STEP_RUN_COMPLETED, STEP_RUN_FAILED,
      STEP_RUN_SKIPPED, STEP_RUN_CANCELED
    ].freeze

    STEP_RUN_TRANSITIONS = {
      [STEP_RUN_PENDING, STEP_RUN_EVENT_WAIT] => STEP_RUN_WAITING,
      [STEP_RUN_PENDING, STEP_RUN_EVENT_START] => STEP_RUN_RUNNING,
      [STEP_RUN_PENDING, STEP_RUN_EVENT_SKIP] => STEP_RUN_SKIPPED,
      [STEP_RUN_PENDING, STEP_RUN_EVENT_CANCEL] => STEP_RUN_CANCELED,
      [STEP_RUN_WAITING, STEP_RUN_EVENT_START] => STEP_RUN_RUNNING,
      [STEP_RUN_WAITING, STEP_RUN_EVENT_CANCEL] => STEP_RUN_CANCELED,
      [STEP_RUN_RUNNING, STEP_RUN_EVENT_COMPLETE] => STEP_RUN_COMPLETED,
      [STEP_RUN_RUNNING, STEP_RUN_EVENT_FAIL] => STEP_RUN_FAILED,
      [STEP_RUN_RUNNING, STEP_RUN_EVENT_CANCEL] => STEP_RUN_CANCELED,
    }.freeze

    def self.can_transition_step_run?(from, event)
      STEP_RUN_TRANSITIONS.key?([from, event])
    end

    def self.transition_step_run(from, event)
      to = STEP_RUN_TRANSITIONS[[from, event]]
      raise ArgumentError, "invalid step run transition: #{from} + #{event}" unless to
      to
    end

    def self.terminal_step_run_status?(status)
      TERMINAL_STEP_RUN_STATUSES.include?(status)
    end
  end
end
